stats/worker.go

```diff
@@ -0,0 +1,375 @@
+		err: make(chan error),
+	}
+	defaultWorker.c <- req
+	if err := <-req.err; err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// GetMeasureByName returns the registered measure associated with name.
+func GetMeasureByName(name string) (Measure, error) {
+	req := &getMeasureByNameReq{
+		name: name,
+		c:    make(chan *getMeasureByNameResp),
+	}
+	defaultWorker.c <- req
+	resp := <-req.c
+	return resp.m, resp.err
+}
+
+// DeleteMeasure deletes an existing measure to allow for creation of a new
+// measure with the same name. It returns an error if the measure cannot be
+// deleted (if one or multiple registered views refer to it).
+func DeleteMeasure(m Measure) error {
+	req := &deleteMeasureReq{
+		m:   m,
+		err: make(chan error),
+	}
+	defaultWorker.c <- req
+	return <-req.err
+}
+
+// GetViewByName returns the registered view associated with this name.
+func GetViewByName(name string) (View, error) {
+	req := &getViewByNameReq{
+		name: name,
+		c:    make(chan *getViewByNameResp),
+	}
+	defaultWorker.c <- req
+	resp := <-req.c
+	return resp.v, resp.err
+}
+
+// RegisterView registers view. It returns an error if the view cannot be
+// registered. Subsequent calls to Record with the same measure as the one in
+// the view will NOT cause the usage to be recorded unless a consumer is
+// subscribed to the view or StartCollectionForAdhoc for this view is called.
+func RegisterView(v View) error {
+	req := &registerViewReq{
+		v:   v,
+		err: make(chan error),
+	}
+	defaultWorker.c <- req
+	return <-req.err
+}
+
+// UnregisterView deletes the previously registered view. It returns an error
+// if the view wasn't registered. All data collected and not reported for the
+// corresponding view will be lost. All clients subscribed to this view are
+// unsubscribed automatically and their subscriptions channels closed.
+func UnregisterView(v View) error {
+	req := &unregisterViewReq{
+		v:   v,
+		err: make(chan error),
+	}
+	defaultWorker.c <- req
+	return <-req.err
+}
+
+// SubscribeToView subscribes a client to a View. If the view wasn't already
+// registered, it will be automatically registered. It allows for many clients
+// to consume the same ViewData with a single registration. -i.e. the aggregate
+// of the collected measurements will be reported to the calling code through
+// channel c. To avoid data loss, clients must ensure that channel sends
+// proceed in a timely manner. The calling code is responsible for using a
+// buffered channel or blocking on the channel waiting for the collected data.
+func SubscribeToView(v View, c chan *ViewData) error {
+	req := &subscribeToViewReq{
+		v:   v,
+		c:   c,
+		err: make(chan error),
+	}
+	defaultWorker.c <- req
+	return <-req.err
+}
+
+// UnsubscribeFromView unsubscribes a previously subscribed channel from the
+// View subscriptions. If no more subscriber for v exists and the the ad hoc
+// collection for this view isn't active, data stops being collected for this
+// view.
```

Comment:
Alternative approach is what google cloud pubsub api uses - it creates a new subscriber that invokes a user provided callback for each received message. And, for unsubscription, it utilizes the context cancellation. I think that allows users to avoid boilerplate for channel creation, loop around channel, etc. and also the API has flexibility to determine what kind of channel is appropriate. What do you think?

Reply:
The issue with that is that if the callback takes too long to run then our worker thread will get stuck and starve the library from executing work.
I guess alternatively we could create a new go routine every time we report the usage but it seems seems wasteful.
I rather give more control to the user and let them decide how they want to process the returned value.